nomad/alloc_endpoint.go

```diff
@@ -565,43 +566,141 @@ func (a *Alloc) SignIdentities(args *structs.AllocIdentitiesRequest, reply *stru
-			reply.Rejections = append(reply.Rejections, &structs.WorkloadIdentityRejection{
-				WorkloadIdentityRequest: *idReq,
-				Reason:                  structs.WIRejectionReasonMissingTask,
-			})
-			continue
-		}
+		job := out.Job
+
+		switch idReq.WorkloadType {
+		case structs.TaskWorkload:
+			task := out.LookupTask(idReq.WorkloadIdentifier)
+			if task == nil {
+				// Job has likely been updated to remove this task
+				reply.Rejections = append(reply.Rejections, &structs.WorkloadIdentityRejection{
+					WorkloadIdentityRequest: *idReq,
+					Reason:                  structs.WIRejectionReasonMissingTask,
+				})
+				continue
+			}
 
-		widFound := false
-		for _, wid := range task.Identities {
-			if wid.Name != idReq.IdentityName {
+			widFound, err := a.signTasks(task, out, idReq, reply, now)
+			if err != nil {
+				return err
+			}
+
+			if !widFound {
+				reply.Rejections = append(reply.Rejections, &structs.WorkloadIdentityRejection{
+					WorkloadIdentityRequest: *idReq,
+					Reason:                  structs.WIRejectionReasonMissingIdentity,
+				})
 				continue
 			}
 
-			widFound = true
-			claims := structs.NewIdentityClaims(out.Job, out, idReq.WorkloadIdentifier, wid, now)
-			token, _, err := a.srv.encrypter.SignClaims(claims)
+		case structs.ServiceWorkload:
+			widFound, err := a.signServices(job, out, idReq, reply, now)
 			if err != nil {
 				return err
 			}
-			reply.SignedIdentities = append(reply.SignedIdentities, &structs.SignedWorkloadIdentity{
-				WorkloadIdentityRequest: *idReq,
-				JWT:                     token,
-				Expiration:              claims.Expiry.Time(),
-			})
-			break
+
+			if !widFound {
+				reply.Rejections = append(reply.Rejections, &structs.WorkloadIdentityRejection{
+					WorkloadIdentityRequest: *idReq,
+					Reason:                  structs.WIRejectionReasonMissingIdentity,
+				})
+				continue
+			}
 		}
+	}
+	return nil
+}
 
-		if !widFound {
-			reply.Rejections = append(reply.Rejections, &structs.WorkloadIdentityRejection{
-				WorkloadIdentityRequest: *idReq,
-				Reason:                  structs.WIRejectionReasonMissingIdentity,
-			})
+func (a *Alloc) signTasks(
+	task *structs.Task,
+	alloc *structs.Allocation,
+	idReq *structs.WorkloadIdentityRequest,
+	reply *structs.AllocIdentitiesResponse,
+	now time.Time,
+) (widFound bool, err error) {
+	for _, wid := range task.Identities {
+		if wid.Name != idReq.IdentityName {
 			continue
 		}
+
+		widFound, err = a.foundAndSignedIdentities(alloc, wid, idReq, reply, now)
+		if err != nil {
+			return
+		}
+		if widFound {
+			return
+		}
 	}
-	return nil
+	return
+}
+
+func (a *Alloc) signServices(
+	job *structs.Job,
+	alloc *structs.Allocation,
+	idReq *structs.WorkloadIdentityRequest,
+	reply *structs.AllocIdentitiesResponse,
+	now time.Time,
+) (widFound bool, err error) {
+	// if it's a task service, we need to figure out what is the task it belongs to
+	// FIXME find a way to do this properly
```

Comment:
I don't understand. `a.SignIdentities` method isn't aware of task or services, and the only information it gets from the incoming request are the identities. In order to verify that the request contains genuine services, the method needs to check what services does the job contain and compare them. Is there a way to figure out what tasks does the job contain? If there is, I don't know it. Otherwise there's no services to call `IdentityHandle` on. 

Or am I missing something?

Reply:
No, you're right that we still have to loop over all the services. But supposing we added a method on `WIHandle` like the following:

```go
func (w WIHandle) Equal(o WIHandle) bool {
	return w.IdentityName == o.IdentityName &&
		w.WorkloadIdentifier == o.WorkloadIdentifier &&
		w.WorkloadType == o.WorkloadType
}
```

Then the `signServices` method could be reduced to this:

```go
func (a *Alloc) signServices(
	job *structs.Job,
	alloc *structs.Allocation,
	idReq *structs.WorkloadIdentityRequest,
	reply *structs.AllocIdentitiesResponse,
	now time.Time,
) (widFound bool, err error) {
	wid := idReq.WIHandle

	// services can be on the level of task groups or tasks
	for _, tg := range job.TaskGroups {
		for _, service := range tg.Services {
			if service.IdentityHandle().Equal(wid) {
				return true, a.signIdentities(alloc, service.Identity, idReq, reply, now)
			}
		}
		for _, task := range tg.Tasks {
			for _, service := range task.Services {
				if service.IdentityHandle().Equal(wid) {
					return true, a.signIdentities(alloc, service.Identity, idReq, reply, now)
				}
			}
		}
	}
	return
}
```

(This example also assumes we change `foundAndSignedIdentities` to just `signIdentities` because we know we've found the identity at that point.)

Reply:
Oooh I get it now, that's good thinking, thanks.

Reply:
addressed in c56f04a25ea6c33a4b84fce80500be710162fd75